### PR TITLE
#15903. Over Disk Quota Paywall

### PR DESCRIPTION
**Pull request by sergiohs84**

### CODE REVIEW

File: include/megaapi.h [hunk 1]
```diff
@@ -6029,6 +6052,11 @@ class MegaGlobalListener
          *     After calling it, this callback will be called again with the corresponding
          *     state if there is really a change.
          *
+         *     - MegaApi::STORAGE_STATE_PAYWALL = 4
+         *     The account has been full for a long time. Now most of actions are disallowed.
+         *     It's needed to call MegaApi::getUserData to check the deadline/warnings
```
Comment (polmr) [inline include/megaapi.h:6057]:
```suggestion
         *     You will need to call MegaApi::getUserData before retrieving the deadline/warnings
```

Comment (polmr) [reply include/megaapi.h:6057]: Plus , we might want to have some words in `getUserData` noting that will also update values such as those

File: src/commands.cpp [hunk 2]
```diff
@@ -452,7 +470,12 @@ void CommandPutFileBackgroundURL::procresult()
 
     if (client->json.isnumeric())
     {
-        error e = (error)client->json.getint();
+        error e = static_cast<error>(client->json.getint());
+        if (e == API_EPAYWALL)
+        {
+            client->activateoverquota(0, true);
+        }
+
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/commands.cpp:478]: Good point! I think the best would be to merge this code with the custom-ETD-error branch, which already copes with the errors in all commands by calling `checkError()`. It is almost ready to merge, so I can update this branch and avoid the repetead block by moving it to the new `checkError()`.
As for the specs, maybe you checked an outdated version? It specifies a long list of commands, not only the `p`: 

> New command-level error code EPAYWALL (name open for debate), likely to be errorcode -29. This must be supported on at least the API commands: m, p, d, g, u, ufa, pfa, pp, ul, l, s, s2, a, dr, dv, rv, ur, ur2, upc, upca, ui, se, sec, usabd, smss, smsv, smsr.

File: src/megaclient.cpp [hunk 3]
```diff
@@ -927,20 +928,24 @@ void MegaClient::activateoverquota(dstime timeleft)
...
     {
         LOG_warn << "Storage overquota";
-        for (transfer_map::iterator it = transfers[PUT].begin(); it != transfers[PUT].end(); it++)
+        int end = (isPaywall) ? PUT : GET;  // in Paywall state, none DLs/UPs can progress
+        for (int d = GET; d <= end; d += PUT - GET)
         {
-            Transfer *t = it->second;
-            t->bt.backoff(NEVER);
-            if (t->slot)
+            for (transfer_map::iterator it = transfers[d].begin(); it != transfers[d].end(); it++)
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaclient.cpp:937]: Yep, definitely. I'll fix it. Thanks

File: src/megaclient.cpp [hunk 4]
```diff
@@ -3217,22 +3223,26 @@ bool MegaClient::abortbackoff(bool includexfers)
     if (includexfers)
     {
         overquotauntil = 0;
-        int end = (ststatus != STORAGE_RED) ? PUT : GET;
-        for (int d = GET; d <= end; d += PUT - GET)
+        if (ststatus != STORAGE_PAYWALL)    // in ODQ Paywall, ULs/DLs are not allowed
         {
-            for (transfer_map::iterator it = transfers[d].begin(); it != transfers[d].end(); it++)
+            // in ODQ Red, only ULs are disallowed
+            int end = (ststatus != STORAGE_RED) ? PUT : GET;
```
[the remark this answers is not in the document]
Comment (sergiohs84) [reply src/megaclient.cpp:3229]: Yes, it makes perfect sense. Let's retry only if state is other than Paywall, no in Paywall.

File: src/commands.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattw-mega) [reply src/commands.cpp:478]: Sounds good - I must have been look at an outdated spec then, thanks

File: include/megaapi.h (hunk 1 above)
Comment (mdsahinco) [reply include/megaapi.h]: Ok

**Comment by sergiohs84** [comment]
trigger compilation

**Comment by joel-meng-nz** [comment]
trigger compilation

File: src/commands.cpp [hunk 5]
```diff
@@ -7929,19 +8020,19 @@ CommandGetRegisteredContacts::CommandGetRegisteredContacts(MegaClient* client, c
...
 void CommandGetRegisteredContacts::procresult()
-{
-    processResult(*client->app, client->json);
-}
-
-void CommandGetRegisteredContacts::processResult(MegaApp& app, JSON& json)
 {
     Error e;
-    if (checkError(e, json))
+    if (checkError(e, client->json))
```
Comment (sergiohs84) [inline src/commands.cpp:8025]: I've just moved the error's handling to the non-static method. The static `processResult()` used by unit-tests only aims to test the JSON response, not the error cases.

File: src/command.cpp [hunk 6]
```diff
@@ -81,16 +84,24 @@ bool Command::checkError(Error& errorDetails, JSON& json)
                        errorDetails.setLinkStatus(json.getint());
                         break;
                     case EOO:
-                        return true;
+                        exit = true;
```
Comment (polmr) [inline src/command.cpp:87]: this now can entail returning false (if !errorDetected). could that be a problem?

File: src/megaapi_impl.cpp [hunk 7]
```diff
@@ -31314,6 +31368,35 @@ void MegaHandleListPrivate::addMegaHandle(MegaHandle h)
...
+
+MegaIntegerList* MegaIntegerListPrivate::copy() const
+{
+    return new MegaIntegerListPrivate(mIntegers);
+}
+
+int64_t MegaIntegerListPrivate::get(int i) const
+{
+    if (i >= static_cast<int>(mIntegers.size()))
+        return -1;
```
Comment (polmr) [inline src/megaapi_impl.cpp:31390]:
```suggestion
        {
                return -1;
        }
```

File: src/command.cpp (hunk 6 above)
Comment (sergiohs84) [reply src/command.cpp:87]: I don't think so. It's made in purpose: if no errocode is read from the JSON (whether `[<error_code>]` or `{"err":<error_code>, ...}`), then there is no error detected at this stage.

File: src/megaapi_impl.cpp (hunk 7 above)
Comment (sergiohs84) [reply src/megaapi_impl.cpp]: Already agreed that `if`s and alike, when followed by single-line, can be left without curly brackets. I also prefer to add them, but not mandatory ;)

**Pull request merged**